### ISSUE TITLE
Home: Current meeting not showing

### ISSUE THREAD

**Issue by shrayshray** [datamade/la-metro-councilmatic#251]
This happened during the 1/18/18 Construction Committee, which began at 10:15. When the issue was reported and I checked the site, it was 11:18; and although the meeting was still running, it wasn't showing in the "Current" slot. I hadn't checked the site to confirm the meeting was indeed up prior to 11:18, but I assume it was based on Regina's explanation:
"Presently, the logic for current meetings is:
- a committee meeting lasts one hour
- a board meeting lasts three hours"
This is in place to account for meetings which occur in succession.

@reginafcompton your suggestion for a fix sounds good: show a committee meeting as "current," until six minutes before next meeting begins. Then, on the days with successive meetings, the "Current meeting" view will display the entire time.
Thank you!

**Comment by reginafcompton** [comment]
@shrayshray - what is the maximum duration of a committee meeting? 

I need to deal with events that occur at the end of a succession of meetings - e.g., [when would the meeting at 1:00 on 1/18 conclude](https://boardagendas.metro.net/event/ad-hoc-customer-experience-committee-aad6c70c626f/) - 90 minutes? less?

(For the others, I am checking the start time of the next meeting...and building out the logic from there.)

**Comment by shrayshray** [comment]
@reginafcompton I think 120 minutes would be safe for maximum meeting duration. 
Does the check for the next meeting just check the next meeting's scheduled time, or does it check whether the next meeting has started, i.e., the live link is available?
If it's possible, it would be ideal to set a 2 hour max, but also have checks for whether the current meeting is still running, and whether the next meeting has started, take precedence for determining when to stop showing the current meeting link.

**Comment by reginafcompton** [comment]
@shrayshray - the logic looks at the start time of the next meeting. For example: This [9:00 am meeting](https://boardagendas.metro.net/event/system-safety-security-and-operations-4a0e05784b62/) would stop showing as current at 10:09am, when [this meeting would show as current](https://boardagendas.metro.net/event/construction-committee-0560d127d6f9/).

**Some other notes and questions**

-  If I remember correctly, the status of an event depends on [the presence of an audio link on the Legistar calendar, during the scheduled meeting time](https://metro.legistar.com/calendar.aspx). _Is that correct?_ 

[ N.B. I though perhaps we could have Councilmatic ping the link, but I do not see a reliable way to programmatically generate it. The link uses a "clip_id," which does not seem to correlate with any ids in Legistar. (For instance, I am looking at the System Safety, Security and Operations Committee meeting on 1/18. It has a ["clip_id" of 772](http://metro.granicus.com/MediaPlayer.php?view_id=2&clip_id=772)).]

- We could also visit the [event detail page on Legistar](https://metro.legistar.com/MeetingDetail.aspx?ID=576613&GUID=13F084ED-0702-4A16-86AD-FE1C7C94BDD1&Options=info&Search=) and check for an audio link there. _A question about this: does the audio link appear on the event detail page, during a current meeting (and then disappear when the meeting ends)?_ 

- How often do meetings overlap? could a 9 am meeting run until 10:22 am, even though another meeting starts at 10:15am?

**Comment by shrayshray** [comment]
@reginafcompton answers to your questions in **bold** below (finally!):

If I remember correctly, the status of an event depends on the presence of an audio link on the Legistar calendar, during the scheduled meeting time. Is that correct?
**Yes.**
[ N.B. I though perhaps we could have Councilmatic ping the link, but I do not see a reliable way to programmatically generate it. The link uses a "clip_id," which does not seem to correlate with any ids in Legistar. (For instance, I am looking at the System Safety, Security and Operations Committee meeting on 1/18. It has a "clip_id" of 772).]
**Actually, "clip_id" is only for links to previously recorded meetings. The live link uses "event_id". See examples here: #238** 

We could also visit the event detail page on Legistar and check for an audio link there. A question about this: does the audio link appear on the event detail page, during a current meeting (and then disappear when the meeting ends)?
**No, it does not. The live link only appears on the Calendar page.**
How often do meetings overlap? could a 9 am meeting run until 10:22 am, even though another meeting starts at 10:15am?
**Really no way to tell how often this will occur.**
**Could the logic be to show the "Listen" link once it's available and leave it up til, let's say, 9pm, unless the link is no longer available on the InSite page (meeting is over), or the next meeting has started?**

**Comment by reginafcompton** [comment]
Thanks for the reply @shrayshray ! 

Here's my thinking: 
The live link only appears on the calendar page. The links posted after a meeting include a `data-event-id`, which correlates with the event GUID (e.g., data-event-id="ac0c87d8-4c4c-4006-a3f1-5cc2ab7a8675"). 

I am going to take a look at these links during today's current meetings. Hopefully, they also include the GUID. If so, then we can hit the calendar page and check for the event GUID of the assumed current meeting (given other derived measures). I think that will give us some more information to help determine if current events overlap.

**Comment by reginafcompton** [comment]
A couple observations:

(1) the "In progress" link _does_ appear on the event detail page. 

<img width="1273" alt="screen shot 2018-02-15 at 11 44 26 am" src="https://user-images.githubusercontent.com/13078679/36271941-a745b72e-1245-11e8-9b64-c40e6a8880fe.png">

(2) meetings can start late? or the event link may not appear until 10, 20, 30 minutes into the meeting?

**Comment by shrayshray** [comment]
(1) Weird! Omar and I checked yesterday during the Planning committee meeting at it wasn't there.
(2) Yes, meetings often start within 10 minutes after the set time. The Board Secretary starts the broadcast/recording only after all board members are seated and ready to begin. It's possible this might be 20 or even 30 minutes after the start time in rare cases.

**Comment by reginafcompton** [comment]
So, we could hit the committee detail page, check for the text "In progress", and if there, assume that the meeting has not ended. 

Something like...
```
import requests
r = requests.get("url-to-committee-detail-page")
if 'In&amp;nbsp;progress' in r.content.decode('utf-8'):
    #  the meeting did not end yet, so keep it as "current" on the index page
```

One important detail:
We do not have valid URLs for the committee detail pages in the OCD API. We need to include those, before implementing this solution.

**Comment by reginafcompton** [comment]
@shrayshray - just to be clear: the absence of the "In progress" link is a **reliable indicator** of when the meeting ends?

**Comment by shrayshray** [comment]
@reginafcompton Sorry for the delay, but needed to confirm absence of "In progress" link is a reliable indicator of when the meeting ends - Yes, it is.
Omar also did some further analysis on average meeting times/max meeting times (see spreadsheet), which indicates looking for a flag like "In progress" will be the most reliable way, as there is an enormous variance on actual meeting duration versus what's scheduled on the calendar.
[Meeting-durations.xlsx](https://github.com/datamade/la-metro-councilmatic/files/1742543/Meeting-durations.xlsx)

**Comment by reginafcompton** [comment]
We will test the effectiveness of the working solution this month, and deploy it to production for April meetings, if all goes well. Below describes the solution we landed on, and alsomemorializes some rejected solutions.

**Working solution: calculate the current meeting, given particular parameters, and use Legistar to determine when the meeting has ended.**

**Rejected solution (though a back up if the working solution does not behave as expected): Determine the end of the current meeting by looking at the start time of the subsequent meeting.**
How would it work? 

**Rejected solution: determine the current meeting by looking at Legistar ONLY**
How would it work? Iterate over all events scheduled within the last six hours, and inspect Legistar for presence of "In progress."
PRO: we can eliminate the complex logic for finding and displaying current meetings. 
PRO: we do not need to worry about the weirdness of "simultaneous meetings."
CON: this solution could add significant load time to the index page, since it has to hit Legistar up to four times. 
CON: the Legistar server can be very slow to respond (again, causing slowness for loading the Metro home page)
CON: the home page may be confusing – meetings can start late (sometimes 30 minutes), and the dissonance between the start time on the web interface and the current-meeting view may confuse users.

```python
if found_events:
    for e in events:
        organization_name = EventParticipant.objects.get(event_id=e.ocd_id).entity_name
        organization_detail_url = Organization.objects.get(name=organization_name).source_url

        r = requests.get(organization_detail_url)
        if 'In&amp;nbsp;progress' in r.content.decode('utf-8'):
            return e
```

**Comment by reginafcompton** [comment]
**Working solution: Use a combination of calculated values and pinging Legistar**
(1) Find all events that occur _five minutes from now_ and in the _past six hours_. (n.b. the maximum recorded meeting duration is 5.38 hours)
(2) Filter those events down to the ones taking place in the last hour. (n.b. the minimum average time a meeting lasts is 52 minutes - hence, one hour....we also need to account for meetings that start late.) If an event is taking place in the last hour, then check if the previous event is marked as "In progress" on Legistar. If so, then display the previous event as current. 
(3) A planned event may not be scheduled to take place in the last hour....but we need to check Legistar for its actual status. Do that. If it's 'In progress' - then, it's current! If not, then return None.

This solution does the following:
- displays an event five minutes before its scheduled start time, unless a previous event overlaps with it (determined by the Legistar "In progress" flag)
- stops displaying an event when (1) an hour has elapsed since the start time, and (2) the Legistar "In progress" flag does not appear. 

CON: we have to hit Legistar when the index page loads - so, slower load time.....maybe we can load part of the page, first? use the partial cache template flags?
CON: if a meeting is shorter than one hour (according to Legistar) it will still render as "Current" on the Councilmatic side.
PRO: the logic for the code is considerably cleaner then previous solutions.

**Comment by reginafcompton** [comment]
**Observations after first day of testing**

(1) The Board Meeting appeared as "Current" as few minutes before its scheduled time - that's what we want!
(2) The meeting stayed "Current," after Legistar added the "In progress" link - again, that's good!
(3) The load time did not seem particularly slow. About 1 or 2 seconds. In this case, Councilmatic only pinged Legistar once. Even so, we should give the request an explicit timeout.
(4) We need to use lxml.html library (or something that gives use html) to parse the page - otherwise, our simple check does not work as expected when the meeting _ends_

**Comment by reginafcompton** [comment]
**Observations during March 14 and 15**

(1) The use of lxml.html fixed number 4 above!
(2) The Finance, Budget, and Audit Committee lasted for less than an hour. So, let's starting pinging Legistar sooner - maybe after 30 minutes?: https://github.com/datamade/la-metro-councilmatic/blob/4f24f74b69dd00ff4f2adb3c39502d35a5f260bf/lametro/utils.py#L41

**Comment by reginafcompton** [comment]
We have a reliable solution in place. To briefly summarize what to expect:

A meeting will show as current five minutes before its scheduled time, even if the "In progress" link does not yet appear on Legistar. The only exception to this: a previous meeting runs long. Then, the on-going meeting will show as current.

A meeting will **stop showing as current** when the "In progress" link disappears. The only exception to this: a meeting lasts less than 45 minutes, which sometimes can happen. (We set this lower limit to minimize the amount of times we hit Legistar.) 

This solution provides a much more precise calculation of when a meeting actually ends (meeting duration can vary greatly, as noted above).

**Comment by reginafcompton** [comment]
Closed by PR #252
